lib/ui/channel_buffers.dart

```diff
@@ -0,0 +1,83 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of dart.ui;
+
+/// A saved platform message for a channel with its callback.
+class StoredMessage {
+  final ByteData _data;
+  final PlatformMessageResponseCallback _callback;
+
+  StoredMessage(this._data, this._callback);
+  ByteData get data => _data;
+  PlatformMessageResponseCallback get callback => _callback;
+}
+
+/// A fixed-size circular queue.
+class RingBuffer<T> {
+  final collection.ListQueue<T> _queue;
+  final int _capacity;
+
+  RingBuffer(this._capacity) 
+    : _queue = collection.ListQueue<T>(_capacity);
+
+  int get length => _queue.length;
+
+  int get capacity => _capacity;
+
+  bool get isEmpty => _queue.isEmpty;
+
+  /// Returns true on overflow.
+  bool push(T val) {
+    bool overflow = false;
+    while (_queue.length >= _capacity) {
+      _queue.removeFirst();
+      overflow = true;
+    }
+    _queue.addLast(val);
+    return overflow;
+  }
+
+  /// Returns null when empty.
+  T pop() {
+    return _queue.isEmpty ? null : _queue.removeFirst();
+  }
+}
+
+/// Storage of channel messages until the channels are completely routed.
+class ChannelBuffers {
+  static const int DEFAULT_BUFFER_SIZE = 100;
+
+  final Map<String, RingBuffer<StoredMessage>> _messages = {};
+
+  void push(String channel, ByteData data, PlatformMessageResponseCallback callback) {
+    RingBuffer<StoredMessage> queue = _messages[channel];
+    if (queue == null) {
+      queue = RingBuffer<StoredMessage>(DEFAULT_BUFFER_SIZE);
+      _messages[channel] = queue;
+    }
+    if (queue.push(StoredMessage(data, callback))) {
+      _Logger._printString('Overflow on channel:' + channel);
```

Comment:
The string is missing a space, and we usually use interpolation: `'Overflow on channel: "$channel".'`

Reply:
We also would expect some kind of "what should you do" explanation for a log. 

The following questions are coming to mind on this one, imagining I'm an end developer:

- Do I need to increase a buffer size somewher?  Can I?
- Am I supposed to be waiting longer for something to happen?  How do I know?  Should I be setting up some sort of communication from Dart before I start pumping messages?
- How do I know if this send originated from my own code? Should I be doing something different if the channel was a Flutter internal channel vs one of my own?

Reply:
Done, resizing is to come at a later PR.  I plan on making it a message.  `Channel('foobar').send("resize 1024", (response) {})`